### ISSUE TITLE
[C++] Reduce number of internal APIs that define default_memory_pool() as default argument value

### ISSUE THREAD

**Issue by felipecrv** [apache/arrow#36360]
### Describe the enhancement requested

Default argument values are convenient, but create implicit uses that are hard to see by simply scanning code at function callsites.

Function that accept a `MemoryPool *` should (ideally) only call functions that perform allocations with that memory pool. Due to many functions being declared with `MemoryPool *pool = default_memory_pool()`, it's very easy to accidentally use the default memory pool when you should be passing down the memory pool you received.

Fixing this issue has two steps:

1) Changing internal functions to not define `default_memory_pool()` as default argument and changing the callsites to pass a memory pool they receive from callers of explicitly pass the `default_memory_pool()`.

2) Change declarations of functions that call `default_memory_pool()` to take a `MemoryPool *` parameter instead. Then go to step 1 again.

Related issues:
 - #36293
 - #34025

### Component(s)

C++

**Comment by felipecrv** [comment]
@wjones127 @pitrou @marin-ma

**Comment by ShaiviAgarwal2** [comment]
@felipecrv 
The enhancement requested is to improve the code by making it easier to see when a function is using the default memory pool instead of the intended memory pool. This is done by changing the internal functions to not define `default_memory_pool()` as the default argument and changing the callsites to explicitly pass the intended memory pool. Additionally, functions that call `default_memory_pool()` should be changed to take a `MemoryPool * `parameter instead.

```
void myFunction(int arg1, int arg2, MemoryPool *pool = default_memory_pool()) {
    // function implementation using pool instead of default_memory_pool()
}

// Updated version:
void myFunction(int arg1, int arg2, MemoryPool *pool) {
    // function implementation using pool instead of default_memory_pool()
}

// Call site before update:
myFunction(arg1, arg2);

// Updated call site:
MemoryPool *pool = getMemoryPool(); // get the intended memory pool
myFunction(arg1, arg2, pool);

```

**Comment by felipecrv** [comment]
> This is done by changing the internal functions to not define default_memory_pool() as the default argument and changing the callsites to explicitly pass the intended memory pool.

Yes, of course. The bad consequence of this is that it breaks APIs that are already public. The multi-step approach I suggested in the description (temporary removal of the default + callsite fixes) would at least ensure the Arrow codebase itself is not using `default_memory_pool()` by accident.

Eventually, we could break some of the APIs and drop the `pool = default_memory_pool()` from parameter declarations.

I'm advocating we always pass `default_memory_pool()` explicitly in Arrow code even when a default value is defined in the declaration. That would make us better judge how annoying being explicit becomes.

**Comment by ShaiviAgarwal2** [comment]
@felipecrv Could you please check whether I'm understanding it correctly or not!!

To enhance the code, we need to modify the internal functions to not define `default_memory_pool()` as a default argument and change the callsites to pass a memory pool they receive from callers or explicitly pass the `default_memory_pool()`. Then we need to change the declarations of functions that call `default_memory_pool()` to take a `MemoryPool * `parameter instead. Finally, we should always pass `default_memory_pool()` explicitly in the Arrow code even when a default value is defined in the declaration.

**Comment by felipecrv** [comment]
@ShaiviAgarwal2 please stop sending ChatGPT-generated comments to the issue tracker.

**Comment by kapoisu** [comment]
Does this issue still require some effort?

Is arrow/ipc/metadata_internal.h a sample that can be changed? I've checked the codebase and most other occurrences are within public headers.
